### PR TITLE
Change handling of boundary conditions

### PR DESCRIPTION
**Pull request by A-CGray**
## Problems
In the static problem class, I noticed that we overwrite the constrained DOF with the correct Dirichlet values inside the static problem's `setVariables` method. This shouldn't be necessary as the boundary conditions are already accounted for in the residual and jacobian such that the solution of the static problem automatically satisfies the BCs. Additionally, overwriting these BC values means our MPhys wrapper is not allowing OpenMDAO to set the states it thinks it is (which has the potential to cause bugs depending on how you're using TACS through MPhys). It also means we cannot perform displacement control for nonlinear solutions, since the dirichlet BCs cannot be applied incrementally. Digging a bit deeper, I found that the reason we do this overwriting is primarily to make OpenMDAO's finite difference partial derivative checks pass for some partial derivatives terms that are not quite computed correctly by TACS as stands:

### addTransposeVectorProduct
- TACS doesn't doesn't have the ability to compute a MatVec product with a transposed `TACSSchurMat` with boundary conditions correctly applied
  - Therefore, the `addTransposeVectorProduct` of the static problem uses the non-transposed matrix, which produces a slightly incorrect result
  - This should result in the MPhys partial derivatives tests failing, but we get around this in a few ways:
    - We use quite a slack tolerance on the mphys partial derivatives tests (`rtol=1e-1`)
    - We overwrite the constrained DOF with the correct Dirichlet values inside the static problem's `setVariables` method. This makes the jacobian computed by OpenMDAO's finite differencing slightly more like the one computed using `addTransposeVectorProduct`
    - Because OpenMDAO tests the norm of the error over the whole jacobian, the test can pass even though some entries related to the BCs are completely wrong

### addSVSens
- The `addSVSens` method of the `TACSAssembler` zeroes out the derivatives of function w.r.t DOFs that are subject to BCs, this is incorrect because the values of those DOFs still affect function values
  - This should mean that the MPhys tests of the function partial derivatives w.r.t states fail.
  - However, overwriting the constrained DOF with the correct Dirichlet values inside the static problem's `setVariables` method means the finite-difference partials with respect the constrained DOF appear to be zero, matching the result of `addSVSens` and thus passing the `check_partials` tests.

## Proposed changes
- [x] Remove call to`setBCs` in `setVariables` so that we're not overwriting the states OpenMDAO thinks it is setting
- [x] Do not zero out boundary condition terms when calling `addSVSens` in static problem
- [x] Add an `applyBCs` argument (that defaults to `true`) to the `TACSAssembler` functions that involve applying BCs:
    - [x] `assembleRes`
    - [x] `assembleJacobian`
    - [x] `assembleMatType`
    - [x] `assembleMatCombo`
    - [x] `addSVSens`
    - [x] `evalMatSVSensInnerProduct`
    - [x] `addJacobianVecProduct`
    - [x] `addMatrixFreeVecProduct`
- [x] Fix `addTransposeJacVecProduct` in static problem by using the `TACSAssembler` `addJacVecProduct` method with the transpose option enabled and the boundary conditions not applied so that it gives the correct result
- [ ] Tighten tolerances in MPhys derivative tests since our partial derivatives should now be correct
- [ ] Implement a scaling factor on the Dirichlet BC residual analagous to the load factor we have for scaling external forces, so that load control and displacement control can be performed using a single scaling factor.

For a more thorough description of the math and why I think this way of handling the BCs is more correct, see [TACS Static Problem Boundary Condition Handling.pdf](https://github.com/user-attachments/files/17601067/TACS.Static.Problem.Boundary.Condition.Handling.pdf)

## Open questions
- How does this BC handling apply to transient and buckling problems?

### CODE REVIEW

**Comment by A-CGray** [comment]
With the exception of `test_thermoelast_linquad_element_2d.py`, I think all the remaining test failures are from transient problems, I think those failures are due to `addSVSens` no longer applying BCs. Will need some input from somebody familiar with the unsteady adjoint implementation to fix that.

**Comment by timryanb** [comment]
Small thing, but can we update the docstrings in `tacs.pyx` to include the new `applyBCs` flag?

**Comment by A-CGray** [comment]
> With the exception of `test_thermoelast_linquad_element_2d.py`, I think all the remaining test failures are from transient problems, I think those failures are due to `addSVSens` no longer applying BCs. Will need some input from somebody familiar with the unsteady adjoint implementation to fix that.

The error in `test_thermoelast_linquad_element_2d` was because the stiffness matrix is non-symmetric in thermoelastic problems. Unlike the static problem, `static_analysis_base_test` actually assembles the transposed matrix for its adjoint solves, which I had undone but now reverted. This transposed matrix is technically not the correct matrix for the adjoint solve because BCs are still applied by zeroing rows, not columns, but it seems to be good enough to pass the tests. Presumably though this means that the current static problem implementation doesn't compute accurate derivatives for thermoelastic problems (we don't test this).

**Comment by A-CGray** [comment]
Not sure what the issue with the unsteady adjoint is now, calling `applyBCs` on the output of `addSVSens` should give exactly the same result as calling it inside `addSVSens` but one of the tests is still failing

**Comment by timryanb** [comment]
> Not sure what the issue with the unsteady adjoint is now, calling `applyBCs` on the output of `addSVSens` should give exactly the same result as calling it inside `addSVSens` but one of the tests is still failing

Should be addressed now

File: tacs/TACS.pyx [hunk 1]
```diff
@@ -2217,14 +2220,15 @@ cdef class Assembler:
...
+        self.ptr.assembleMatType(matType, A.ptr, matOr, loadScale, applyBCs)
         return
 
     def assembleMatCombo(self, ElementMatrixType matType1, double scale1,
                          ElementMatrixType matType2, double scale2, Mat A,
                          MatrixOrientation matOr=NORMAL,
-                         TacsScalar loadScale=1.0):
+                         TacsScalar loadScale=1.0, bool applyBCs=True):
         """
         Assemble a combination of two matrices
```
[the remark this answers is not in the document]
Comment (A-CGray) [reply tacs/TACS.pyx]: Done

**Review by timryanb — APPROVED** [review]
